gu-provider/src/dockerman.rs

```diff
@@ -106,16 +106,31 @@ impl DockerSession {
             config
         };
 
+        let container_copy = self.container.clone();
+
         self.container
             .exec(&cfg)
             .map_err(|e| format!("{}", e))
-            .fold(String::new(), |mut s, (_t, it)| {
-                match std::str::from_utf8(it.into_bytes().as_ref()) {
-                    Ok(chunk_str) => s.push_str(chunk_str),
-                    Err(_) => (),
-                };
-
-                Ok::<String, String>(s)
+            .and_then(|(stream, id)| {
+                stream
+                    .fold(String::new(), |mut s, (_t, it)| {
+                        match std::str::from_utf8(it.into_bytes().as_ref()) {
+                            Ok(chunk_str) => s.push_str(chunk_str),
+                            Err(_) => (),
+                        };
+                        Ok::<String, String>(s)
+                    })
+                    .map_err(|e| format!("{}", e))
+                    .and_then(move |output| {
+                        let exit_status = container_copy
+                            .check_exec_status(&id)
+                            .map_err(|e| format!("{}", e));
+                        exit_status.join(future::ok(output))
+                    })
+                    .and_then(|(status, output)| match status {
+                        0 => future::ok(output),
+                        exit_code => future::err(format!("{}\n{}", output, exit_code)),
```

Comment:
```suggestion
                        exit_code => future::err(format!("{}\nExit code: {}", output, exit_code)),
```
We can't really do much better than this until #240 is addressed, but we should be a little more informative to the end-user.